arrow-syntax/src/main/kotlin/arrow/syntax/function/memoization.kt

```diff
@@ -214,11 +214,9 @@ private data class MemoizeKey22<out P1, out P2, out P3, out P4, out P5, out P6,
 }
 
 private class MemoizedHandler<F, in K : MemoizedCall<F, R>, out R>(val f: F) {
-    private val m = ConcurrentHashMap<K, R>()
+    private val m = newConcurrentMap<K, R>()
     operator fun invoke(k: K): R = m[k] ?: run({
-        val r = k(f)
-        m.putIfAbsent(k, r)
-        r
+        m[k].let { it ?: k(f).also { m[k] = it } }
```

Comment:
@MarioArias I had to change the logic not to use `putIfAbsent`. This gets a nullable, and if it doesn't exist it executes the function and returns it. Is it the case that the function should always be called?

Reply:
We lost thread safety which is of paramount importance in this case; we must use `putIfAbsent` or expose any other equivalent function via a proxy

Reply:
Okay, lemme proxy it!

Reply:
It's even cooler now :D